### PR TITLE
Added a new option : -n/--no-window

### PR DESCRIPTION
**Pull request by TheCapsLock**
This option allows scrcpy to be run withou showing preview window : we can now use
scrcpy only for screen recording as an example

### CODE REVIEW

File: app/src/scrcpy.c [hunk 1]
```diff
@@ -80,7 +82,9 @@ static SDL_bool event_loop(void) {
                 if (!screen.has_frame) {
                     screen.has_frame = SDL_TRUE;
                     // this is the very first frame, show the window
-                    screen_show_window(&screen);
+                    if (!no_window) {
+                        screen_show_window(&screen);
+                    }
```
Comment (rom1v) [inline app/src/scrcpy.c]: I think you can skip code far earlier, you may not even need to initialize SDL nor decode frames at all if you have no window.

File: app/src/main.c [hunk 2]
```diff
@@ -209,6 +215,7 @@ static SDL_bool parse_args(struct args *args, int argc, char *argv[]) {
         {"bit-rate",     required_argument, NULL, 'b'},
         {"crop",         required_argument, NULL, 'c'},
         {"fullscreen",   no_argument,       NULL, 'f'},
+        {"no-window",   no_argument,       NULL, 'n'},
```
Comment (rom1v) [inline app/src/main.c]: alignment

File: app/src/main.c [hunk 3]
```diff
@@ -219,7 +226,7 @@ static SDL_bool parse_args(struct args *args, int argc, char *argv[]) {
         {NULL,           0,                 NULL, 0  },
     };
     int c;
-    while ((c = getopt_long(argc, argv, "b:c:fhm:p:r:s:tv", long_options, NULL)) != -1) {
+    while ((c = getopt_long(argc, argv, "b:c:fnhm:p:r:s:tv", long_options, NULL)) != -1) {
```
Comment (rom1v) [inline app/src/main.c]: (please keep the alphabetical order, ditto above and below)

**Comment by npes87184** [comment]
Hi,

I have two simple recommendations for this patch:
* Please provide either a log in console or a toast in android side, so that we can know that the scrcpy finishes its initialization.
* Use -n alone doesn’t make sense now. IMO, we can return error if -n and -r don’t exist simultaneously.

Thanks,
Yu-Chen

File: app/src/main.c [hunk 4]
```diff
@@ -232,6 +239,9 @@ static SDL_bool parse_args(struct args *args, int argc, char *argv[]) {
             case 'f':
                 args->fullscreen = SDL_TRUE;
                 break;
+            case 'n':
+                args->no_window = SDL_TRUE;
```
Comment (npes87184) [inline app/src/main.c]: Missing default value or an initialization. See [main.c](
https://github.com/Genymobile/scrcpy/blob/8d4ff03b7feab7123dd5483f16ff282fddbb425d/app/src/main.c#L291)

Comment (TheCapsLock) [reply app/src/main.c]: done

File: app/src/main.c (hunk 3 above)
Comment (TheCapsLock) [reply app/src/main.c]: done

File: app/src/main.c (hunk 2 above)
Comment (TheCapsLock) [reply app/src/main.c]: done; sorry for that

**Comment by TheCapsLock** [comment]
>     * Use -n alone doesn’t make sense now. IMO, we can return error if -n and -r don’t exist simultaneously.

Added an error message if nor -n is provided without -r

**Comment by TheCapsLock** [comment]
> Thank you for your contribution.
> 
> I like the idea to add such a flag. Maybe it could skip a lot more code (do not decode frame, do not initialize SDL…
> 
> I'll test it probably after the FOSDEM.

Thanks a lot for that ; I tried some combinations but did not get something better yet ; if you have clear idea in mind may I let you look at this specific topic ?

**Comment by TheCapsLock** [comment]
>     * Please provide either a log in console or a toast in android side, so that we can know that the scrcpy finishes its initialization.

I've added a log message, thank you for this clever recommendation

**Comment by rom1v** [comment]
> if you have clear idea in mind may I let you look at this specific topic ?

Yes, it requires some refactor beforehand: the decoder must be separated from the "video stream receiver", so that we can skip the decoder completely (we don't need to decode the video at all if there is no window). I will try to work on it so that it is included in the next version.

**Comment by TheCapsLock** [comment]
> > if you have clear idea in mind may I let you look at this specific topic ?
> 
> Yes, it requires some refactor beforehand: the decoder must be separated from the "video stream receiver", so that we can skip the decoder completely (we don't need to decode the video at all if there is no window). I will try to work on it so that it is included in the next version.

Great to here this kind of news :) thanks a lot!

**Comment by rom1v** [comment]
I refactored to split the decoder into stream+decoder: https://github.com/Genymobile/scrcpy/commit/ce5cdd877dfb9d195d0afd5be1a052b982a12a70.
Then I rebased your commit (but removed the part which disabled the window after decoding): https://github.com/Genymobile/scrcpy/commit/dbf51f15ac8ef0bae73179482ca28827aef5923b.
Then I disabled everything useless when `--no-window` is passed: https://github.com/Genymobile/scrcpy/commit/8c95835de9b05c0cee91f79016a84614ff82e226.

It's on `dev` branch for now.

I'm closing as _merged_ (even if github does see it as merged).

**Comment by rom1v** [comment]
I have renamed the option `-N`/`--no-display`.

The description of scrcpy is "Display and control your Android device".

We want an option to disable display, another one to disable control, so I also implemented `-n`/`--no-control`.

**Comment by TheCapsLock** [comment]
Tested on multiple devices, seems to work great, thanks :heart:

**Comment by Gapplay** [comment]
@rom1v is possible to show fps when No-display is enabled?

**Comment by rom1v** [comment]
Currently, no. 

A command line option (`--print-fps`) could be added to enable it on start.

**Comment by rom1v** [comment]
@Gapplay Please open a new issue for it, to not forget it.

**Pull request closed without merging**